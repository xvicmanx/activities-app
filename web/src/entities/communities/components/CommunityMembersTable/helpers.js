// @flow

export const coordinatesTableValueResolver = (item: Object): string =>
  item.coordinates ? 'Si' : 'No';
