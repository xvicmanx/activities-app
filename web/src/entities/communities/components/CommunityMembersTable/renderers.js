// @flow

import React from 'react';

import UsersDropdown from '../../../users/components/UsersDropdown';
import BooleanDropdown from '../BooleanDropdown';

type RendererProps = {
  field: Object,
};

export const MembersSelectRenderer = ({ field }: RendererProps): React$Element<any> => (
  <UsersDropdown {...field} />
);

export const CoordinatesSelectRenderer = ({ field }: RendererProps): React$Element<any> => (
  <BooleanDropdown {...field} />
);
