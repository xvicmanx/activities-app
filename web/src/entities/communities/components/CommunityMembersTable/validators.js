// @flow

export const addMemberFormValidator = (members: Array<Object>): ((values: any) => any) => (values: Object): Object => {
  const errors = {};

  if (!values.memberId) {
    errors.memberId = 'Por favor, seleccione miembro';
  }

  if (members.map((x) => x.id).includes(values.memberId)) {
    errors.memberId =
      'El miembro seleccionado ya existe en la comunidad';
  }

  return errors;
};
