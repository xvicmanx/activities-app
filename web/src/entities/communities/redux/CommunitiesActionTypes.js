export const SET_COMMUNITIES = 'SET_COMMUNITIES';
export const SET_COMMUNITIES_LOADING_STATE = 'SET_COMMUNITIES_LOADING_STATE';
export const SET_COMMUNITIES_ERROR = 'SET_COMMUNITIES_ERROR';

export const SET_COMMUNITY = 'SET_COMMUNITY';
export const SET_COMMUNITY_LOADING_STATE = 'SET_COMMUNITY_LOADING_STATE';
export const SET_COMMUNITY_ERROR = 'SET_COMMUNITY_ERROR';

export const ADD_MEMBER = 'ADD_MEMBER';
