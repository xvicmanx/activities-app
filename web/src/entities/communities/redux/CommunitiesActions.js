import {
  SET_COMMUNITIES,
  SET_COMMUNITIES_LOADING_STATE,
  SET_COMMUNITIES_ERROR,
  SET_COMMUNITY,
  SET_COMMUNITY_LOADING_STATE,
  SET_COMMUNITY_ERROR,
  ADD_MEMBER,
} from './CommunitiesActionTypes';

import CommunitiesService from '../services/CommunitiesService';

export const fetchCommunities = (token) => async (dispatch) => {
  dispatch({
    type: SET_COMMUNITIES_LOADING_STATE,
    payload: true,
  });

  const res = await CommunitiesService.fetchCommunities(token);

  if (res.success) {
    dispatch({
      type: SET_COMMUNITIES,
      payload: res.communities,
    });
  } else {
    dispatch({
      type: SET_COMMUNITIES_ERROR,
      payload: res.message,
    });
  }

  dispatch({
    type: SET_COMMUNITIES_LOADING_STATE,
    payload: false,
  });
};


export const fetchCommunity = (token, id) => async (dispatch) => {
  dispatch({
    type: SET_COMMUNITY_LOADING_STATE,
    payload: true,
  });

  const res = await CommunitiesService.fetchCommunity(token, id);

  if (res.success) {
    dispatch({
      type: SET_COMMUNITY,
      payload: res.community,
    });
  } else {
    dispatch({
      type: SET_COMMUNITY_ERROR,
      payload: res.message,
    });
  }

  dispatch({
    type: SET_COMMUNITY_LOADING_STATE,
    payload: false,
  });
};


export const addMember = (token, communityId, memberId, coordinates) => async (dispatch) => {
  dispatch({
    type: SET_COMMUNITY_LOADING_STATE,
    payload: true,
  });

  const res = await CommunitiesService.addMember(token, communityId, memberId, coordinates);

  if (res && res.success) {
    dispatch({
      type: ADD_MEMBER,
      payload: res.member,
    });
  } else {
    dispatch({
      type: SET_COMMUNITY_ERROR,
      payload: res ? res.message : 'Unexpected error',
    });
  }

  dispatch({
    type: SET_COMMUNITY_LOADING_STATE,
    payload: false,
  });
};