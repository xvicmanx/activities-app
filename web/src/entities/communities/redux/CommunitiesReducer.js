import {
  SET_COMMUNITIES,
  SET_COMMUNITIES_LOADING_STATE,
  SET_COMMUNITIES_ERROR,
  SET_COMMUNITY,
  SET_COMMUNITY_LOADING_STATE,
  SET_COMMUNITY_ERROR,
  ADD_MEMBER,
} from './CommunitiesActionTypes';

const INITIAL_STATE = {
  communities: null,
  community: null,
  member: null,
  loading: false,
  error: null,
};

const CommunitiesReducer = (state = INITIAL_STATE, { type, payload }) => {
  switch (type) {
    case SET_COMMUNITIES_LOADING_STATE:
      return {
        ...state,
        loading: payload,
        error: payload ? null : state.error,
      };

    case SET_COMMUNITIES_ERROR:
      return {
        ...state,
        error: payload,
      };

    case SET_COMMUNITIES:
      return {
        ...state,
        communities: payload,
      };

    case ADD_MEMBER:
      return {
        ...state,
        community: {
          ...state.community,
          members: [...(state.community.members || []), payload],
        },
      };

    case SET_COMMUNITY_LOADING_STATE:
      return {
        ...state,
        loading: payload,
        error: payload ? null : state.error,
      };

    case SET_COMMUNITY_ERROR:
      return {
        ...state,
        error: payload,
      };

    case SET_COMMUNITY:
      return {
        ...state,
        community: payload,
      };

    default:
      return state;
  }
};

export default CommunitiesReducer;
