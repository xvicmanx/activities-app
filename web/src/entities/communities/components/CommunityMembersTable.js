import React from 'react';
import { Link } from 'react-router-dom';
import { useDispatch } from 'react-redux';
import CRUDTable, { CreateForm, DeleteForm, Fields, Field } from 'react-crud-table';

import { readTokenFromCookie } from '../../users/redux/UsersActions';
import { addMember } from '../redux/CommunitiesActions';
import UsersDropdown from '../../users/components/UsersDropdown';
import BooleanDropdown from './BooleanDropdown';

const styles = {
  container: {
    margin: "auto",
    width: "fit-content",
  },
};

const MembersSelectRenderer = ({ field }) => (
  <UsersDropdown {...field}/>
);

const CoordinatesSelectRenderer = ({ field }) => (
  <BooleanDropdown {...field} />
);


const CommunityMembersTable = ({ communityId, members }) => {
  const dispatch = useDispatch();
  const token = readTokenFromCookie();
  return (
    <CRUDTable
      caption="Miembros"
      items={members}
    >
      <Fields>
        <Field
          name="memberId"
          label="Miembro"
          hideFromTable
          render={MembersSelectRenderer}
        />
        <Field
          name="name"
          label="Nombre"
          placeholder="Nombre"
          hideInCreateForm
        />
        <Field
          name="coordinates"
          label="Coordinador?"
          tableValueResolver={item => item.coordinates ? 'Si' : 'No'}
          render={CoordinatesSelectRenderer}
        />
      </Fields>
      <CreateForm
        title="Agregar miembro"
        message="Agregar miembro"
        trigger="Agregar miembro"
        onSubmit={item => dispatch(addMember(
          token,
          communityId,
          item.memberId,
          item.coordinates,
        ))}
        submitText="Agregar"
        validate={(values) => {
          const errors = {};
          
          if (!values.memberId) {
            errors.memberId = 'Por favor, seleccione miembro';
          }

          if (members.map(x => x.id).includes(values.memberId)) {
            errors.memberId = 'El miembro seleccionado ya existe en la comunidad';
          }
  
          return errors;
        }}
      />
    </CRUDTable>
  );
};

export default CommunityMembersTable;
