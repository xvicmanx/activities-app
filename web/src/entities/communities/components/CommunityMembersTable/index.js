// @flow

import React from 'react';
import { useDispatch } from 'react-redux';
import CRUDTable, { CreateForm, Fields, Field } from 'react-crud-table';

import { readTokenFromCookie } from '../../../users/redux/UsersActions';
import { addMember } from '../../redux/CommunitiesActions';
import { CoordinatesSelectRenderer, MembersSelectRenderer } from './renderers';
import { addMemberFormValidator } from './validators';
import { coordinatesTableValueResolver } from './helpers';


type Props = {
  communityId: number,
  members: Array<{ id: number | string }>,
};

const CommunityMembersTable = ({
  communityId,
  members,
}: Props): React$Element<any> => {
  const dispatch = useDispatch();
  const token = readTokenFromCookie();
  return (
    <CRUDTable caption="Miembros" items={members}>
      <Fields>
        <Field
          name="memberId"
          label="Miembro"
          hideFromTable
          render={MembersSelectRenderer}
          sortable={false}
        />
        <Field
          name="name"
          label="Nombre"
          placeholder="Nombre"
          hideInCreateForm
          sortable={false}
        />
        <Field
          name="coordinates"
          label="Coordinador?"
          tableValueResolver={coordinatesTableValueResolver}
          render={CoordinatesSelectRenderer}
          sortable={false}
        />
      </Fields>
      <CreateForm
        title="Agregar miembro"
        message="Agregar miembro"
        trigger="Agregar miembro"
        onSubmit={(item) =>
          dispatch(
            addMember(token, communityId, item.memberId, item.coordinates)
          )
        }
        submitText="Agregar"
        validate={addMemberFormValidator}
      />
    </CRUDTable>
  );
};

export default CommunityMembersTable;
