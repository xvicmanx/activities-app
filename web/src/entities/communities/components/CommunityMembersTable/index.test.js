import React from 'react';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { act } from 'react-dom/test-utils';

import CommunityMembersTable from './';

jest.mock('react-redux');
jest.mock('react-toastify');
jest.mock('../../redux/CommunitiesActions');
jest.mock('./renderers', () => ({
  CoordinatesSelectRenderer: () => 'CoordinatesSelectRenderer',
  MembersSelectRenderer: () => 'MembersSelectRenderer',
}));

import { useDispatch } from 'react-redux';
import { toast } from 'react-toastify';
import { addMember } from '../../redux/CommunitiesActions';

Enzyme.configure({ adapter: new Adapter() });

describe('CommunityMembersTable', () => {
  const members = [
    {
      id: 1,
      name: 'Test member 1',
    },
    {
      id: 2,
      name: 'Test member 2',
    },
  ];
  const dispatch = jest.fn();

  beforeEach(() => {
    useDispatch.mockClear().mockImplementation(() => dispatch);
    addMember.mockClear().mockImplementation(() => ({ addMemberResult: true }));
    dispatch.mockClear();
  });

  it('renders properly', async () => {
    const result = Enzyme.mount(
      <CommunityMembersTable members={members} communityId={1} />
    );
    expect(result.find('CRUDTable').props()).toMatchSnapshot();
  });

  it('adds member on create form submit', async () => {
    const result = Enzyme.shallow(
      <CommunityMembersTable members={members} communityId={1} />
    );

    // Access create form
    result.childAt(1).props().onSubmit({
      memberId: 3,
      coordinates: true,
    });
    result.update();

    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(dispatch).toHaveBeenCalledWith({ addMemberResult: true });

    expect(toast.success).toHaveBeenCalledTimes(1);
    expect(toast.success).toHaveBeenCalledWith(
      'Miembro agregado de manera exitosa!'
    );
  });
});
