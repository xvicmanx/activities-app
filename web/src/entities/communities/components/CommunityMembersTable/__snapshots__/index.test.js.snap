// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommunityMembersTable renders properly 1`] = `
Object {
  "actionsLabel": "Actions",
  "caption": "Miembros",
  "children": Array [
    <CRUDTable_Fields>
      <CRUDTable_Field
        hideFromTable={false}
        hideInCreateForm={true}
        hideInUpdateForm={true}
        label="Id"
        name="id"
        queryable={true}
        sortable={false}
        tableValueResolver={null}
        type="text"
      />
      <CRUDTable_Field
        hideFromTable={true}
        hideInCreateForm={false}
        hideInUpdateForm={false}
        label="Miembro"
        name="memberId"
        queryable={true}
        render={[Function]}
        sortable={false}
        tableValueResolver={null}
        type="text"
      />
      <CRUDTable_Field
        hideFromTable={false}
        hideInCreateForm={true}
        hideInUpdateForm={false}
        label="Nombre"
        name="name"
        placeholder="Nombre"
        queryable={true}
        sortable={false}
        tableValueResolver={null}
        type="text"
      />
      <CRUDTable_Field
        hideFromTable={false}
        hideInCreateForm={false}
        hideInUpdateForm={false}
        label="Coordinador?"
        name="coordinates"
        queryable={true}
        render={[Function]}
        sortable={false}
        tableValueResolver={[Function]}
        type="text"
      />
    </CRUDTable_Fields>,
    <CRUDTable_CreateForm
      message="Agregar miembro"
      onSubmit={[Function]}
      submitText="Agregar"
      title="Agregar miembro"
      trigger="Agregar miembro"
      validate={[Function]}
    />,
  ],
  "fetchItems": null,
  "items": Array [
    Object {
      "id": 1,
      "name": "Test member 1",
    },
    Object {
      "id": 2,
      "name": "Test member 2",
    },
  ],
  "onChange": [Function],
  "showQueryBuilder": false,
}
`;
